components/service/fretboard/src/main/java/mozilla/components/service/fretboard/source/kinto/SignatureVerifier.kt

```diff
@@ -0,0 +1,266 @@
+     */
+    private fun getX5U(url: URL): PublicKey {
+        val certs = ArrayList<X509Certificate>()
+        val cf = CertificateFactory.getInstance("X.509")
+        val response = client.get(url)
+        val reader = BufferedReader(StringReader(response))
+        val firstLine = reader.readLine()
+        if (firstLine != "-----BEGIN CERTIFICATE-----")
+            throw ExperimentDownloadException("")
+        var certPem = firstLine
+        certPem += '\n'
+
+        reader.readLines().forEach {
+            certPem += it
+            certPem += '\n'
+            if (it == "-----END CERTIFICATE-----") {
+                val cert = cf.generateCertificate(ByteArrayInputStream(certPem.toByteArray()))
+                certs.add(cert as X509Certificate)
+                certPem = ""
+            }
+        }
+        if (certs.count() < MIN_CERTIFICATES) {
+            throw ExperimentDownloadException("The chain must contain at least 2 certificates")
+        }
+        verifyCertChain(certs)
+        return certs[0].publicKey
+    }
+
+    /**
+     * Verifies the validity of the certificates on the chain
+     *
+     * @param certificates certificates
+     *
+     * @throws ExperimentDownloadException if any certificate is not valid
+     */
+    private fun verifyCertChain(certificates: List<X509Certificate>) {
+        for (i in 0 until certificates.count()) {
+            val cert = certificates[i]
+            if (!isCertValid(cert)) {
+                throw ExperimentDownloadException("Certificate expired or not yet valid")
+            }
+            if ((i + 1) == certificates.count()) {
+                verifyRoot(cert)
+            } else {
+                verifyCertSignedByParent(cert, i, certificates)
+            }
+        }
+    }
+
+    /**
+     * Verifies that the certificate is signed by its parent on the chain
+     *
+     * @param certificate certificate
+     * @param index index of the certificate on the chain
+     * @param certificates certificate chain
+     *
+     * @throws ExperimentDownloadException if the certificate chain is invalid
+     *
+     * @return true if the certificate is signed by its parent on the chain, false otherwise
+     */
+    private fun verifyCertSignedByParent(
+        certificate: X509Certificate,
+        index: Int,
+        certificates: List<X509Certificate>
+    ) {
+        var signed = true
+        try {
+            certificate.verify(certificates[index + 1].publicKey)
+        } catch (e: CertificateException) {
+            signed = false
+        } catch (e: NoSuchAlgorithmException) {
+            signed = false
+        } catch (e: InvalidKeyException) {
+            signed = false
+        } catch (e: NoSuchProviderException) {
+            signed = false
+        } catch (e: SignatureException) {
+            signed = false
+        }
+
+        if (!signed) {
+            throw ExperimentDownloadException("Invalid certificate chain")
+        }
```

Comment:
Now that `ExperimentDownloadException` can wrap an exception I think we want to do this here. And I assume we can do that in the catch blocks directly and do not need `signed` here anymore?

Reply:
I have changed that now :)